### ISSUE TITLE
Newly created Daily's Start Date should be yesterday if the time of creation is between midnight and CDS

### ISSUE THREAD

**Issue by hairlessbear** [HabitRPG/habitica#5525]
_edit by admin:_

_Currently, when a Daily is created, the start date is set to the current real-world date (i.e., the date that your PC's clock reports)._ 
_However if you have a Custom Day Start (CDS) set and if you create the Daily after midnight but before the CDS, then the start date should be set to yesterday (because HabitRPG-time is still "yesterday" until the CDS time is reached). This allows the new Daily to be active as soon as it is created._

**Suggested fix**:

When a Daily is created, get the current hour from the date object. If the hour is lower than the Custom Day Start, subtract 24 hours from the date and then set it as the startDate. Here's how it's already done in the iOS app:
https://github.com/HabitRPG/habitrpg-ios/blob/master/HabitRPG/TableViewController/Tasks/HRPGFormViewController.m#L119 

(Thanks @vIiRuS)

---

**Edit by hairlessbear**: for context, the original title of this issue was "Newly created dailies are automatically grey, but are still due every day". Original post is below.

---

[Video of the behavior](https://vid.me/ty0q)

In the video, a tag is selected when the daily is created, but the behavior occurred regardless of tag selection. It also occurred after reloading the page, as well as logging out and back in again.

In the DDT, the new "grey" dailies don't actually appear to be grey. They are categorized as normal dailies, so I suspect this is just a display problem.

If it helps, UUID is 8ac7242d-8f18-4d6f-b4f4-8f6302f7fa0b. I noticed this behavior occurring at around 00:30, and my cron is at 1.

**Comment by Alys** [comment]
That's probably because of the start date. I am guessing that the start date on that Daily is "today" in real terms, i.e., the current calendar date. However because of your CDS and the current time, in HabitRPG you are still on yesterday. Therefore the Daily is not due at the moment, but it will be after 1am.

We're taking this approach because the start date field shows just the date and doesn't give any ability to change the time portion of that date.

I can explain further if this isn't clear.

**Comment by Alys** [comment]
Also, we're pretending that I didn't type "you're" instead of "your". I do English good.

I'm still looking at the Start Date code to see if I can make it better. It is unintuitive in some circumstances and possibly broken in others.

**Comment by hairlessbear** [comment]
It seems like a daily should be due on the day it is created (i.e., before next cron). Not only that, within the daily's display (and in the DDT), it _is_ due today.

**Comment by hairlessbear** [comment]
At least, I think that's true. I'm creating a daily that I'm going to leave unchecked, as a test.

**Comment by Alys** [comment]
"within the daily's display, it is due today"  - Can you explain what you mean by that? You said it was grey, which means the display shows that it's not due.

**Comment by hairlessbear** [comment]
I mean that each day of the week is...I don't know the right word..."selected"? The visual style of each day is the style that indicates that a daily is due that day (bolder outline and slightly darker)

**Comment by Alys** [comment]
Please look at the start date. That takes precedence over all the other due/not due settings. If the start date is in the future, then it will never be due.

If the start date is today's actual date but you are still before your CDS time, then your HabitRPG-day is still yesterday, so the Daily is not due at this precise moment.

**Comment by hairlessbear** [comment]
That makes sense. In that case, the display is extremely unclear. The grey indicates it's not due (which is correct), but the style of the days of the week indicate that it is due, making it appear that something is wrong (hence why I filed this bug report :smile:)

**Comment by hairlessbear** [comment]
Just ran cron, and I _think_ the test daily did damage me, indicating that it was actually due. I don't know of a method to test this with certainty though. I believe it was ticked off because it now has a history (i.e., a blue dot) when you click on the "progress" button, whereas a new daily does nothing when the button is pressed.

**Comment by Alys** [comment]
Not exactly. The style of the days of the week indicate which days of the week you have selected as the days it should be due on. We need to have that style still applied to the days even when the start date is in the future because otherwise it would be harder to set up a Daily with all the correct days in advance of the start date.

**Comment by Alys** [comment]
It will have a history even if it didn't damage you. The history can indicate no change in value (i.e, if the Daily wasn't due, then the new history value will be the same as the old one).

**Comment by hairlessbear** [comment]
Good to know about the history.

Regarding the visual style, the issue I have is that if all days of the week are selected, and the daily is unchecked, it's very unintuitive for it to be grey. In all other situations, an unchecked daily with all days of the week selected would not display as grey, and the fact that it does in this one context is confusing.

**Comment by Alys** [comment]
But its grey because it's **not** due because the start date is in the future.

If you take any existing Daily and change its start date to the future, the Daily has to go grey because it's no longer due, regardless of any other settings it might have. 

We can't adjust the appearance of the days of the week boxes when the start date changes, because those days haven't been changed. If we did make it look as if the days of the week had been unselected, that would be a bug because it wouldn't be correctly showing the actual settings of the Daily.

You are right that what you are seeing now would have been a bug a few weeks ago - i.e., if the days of the week were all selected but the Daily was grey, that would have been wrong then. However the situation is different now because the start date can be changed to be a future date.

**Comment by hairlessbear** [comment]
I understand that it's not a bug (EDIT: to clarify, I understand it, _now that you've explained it to me_ :smile:). What I'm saying is that it _feels_ like a bug; the behavior is unintuitive, and if possible, should be changed. It doesn't make sense to me to have dailies I create at one time of my "HabitRPG day" to be due before cron, and have others not be. The whole point of CDS (as I understand it) is to allow users to base the time they can perform all normal HabitRPG functions on their own schedule, not when midnight hits. Having features that change behavior based on midnight goes directly against CDS's purpose.

What if the code that determines the default start date used the CDS? If the CDS is after midnight, and the daily is being created before the CDS, it could set the start date to the day before, making the daily due on the "HabitRPG day" it's created. That would mean that all dailies created before cron are due before cron by default (which is what the behavior was before this feature was added), but a user can set them to be due in the future if they choose (by the way, the ability to change dailies' start dates truly is awesome, and I'm _so_ glad you guys added it in).

**Comment by Alys** [comment]
Okay but that's a different thing than changing the days of the week appearance to make them look as if they haven't been selected.

You are right about the point of CDS - it does let you perform actions for a day up until a certain time on the next day. And you CAN add a new Daily between midnight and your CDS and still have it due immediately - you just have to adjust the start date to be "yesterday". Remember that if your PC's date is, for example, 30 June, but you have a CDS of 1am and the current time is 00:30, then the HabitRPG day is actually 29 June. HabitRPG-day becomes 30 June only after the CDS time has passed.

Therefore if you're creating a new Daily between midnight and your CDS time and you want it due immediately, change its start date to 29 June - because that's the actual date now as far as HabitRPG is concerned (for your account, because of your CDS).

The sole problem with the Daily you created before 1am "today" is that the start date WAS in the future as far as your HabitRPG-day was concerned.

We can consider different default start dates for tasks, and in fact I'm working on the start date code right now to try to improve it, but until that's done, you do need to look at the start date that's been set in the Daily and remember that it is considered as being in the future if it's today's real date but if your HabitRPG day is still yesterday.

**Comment by hairlessbear** [comment]
I was unclear before: I'm not suggesting that the visual style should be modified to make days appear as not selected when they really are. In fact, I'm not suggesting that the visual style be modified in any way. I can see why it appeared that I was suggesting visual changes; I didn't mean to give that impression, and should have clarified. 

What I'm suggesting is that the daily creation behavior should be as it was before the start date functionality was added (i.e., dailies are due on the same "HabitRPG day" they are created), because that's more intuitive than the current behavior. As far as I can tell, the best way to do this would be to have the default due date be based on the CDS, but it sounds like you're already working on something like that.

As always, thanks for taking the time to explain/discuss the code with me. Much appreciated.

I'll go ahead and close this since it's not a bug.

**Comment by Alys** [comment]
Actually I'll reopen this and adjust the title a bit, because you're right that the initial start date should take into account the CDS time and it's probably not something I'll get to today but I don't want to forget it. I'm sorry it was confusing; you were right to report it.

**Comment by nasfarley88** [comment]
Does this still need to be done?

I've been looking at this issue and I'm not sure it's as trivial to fix as starting it the next day. E.g.

My CDR is 3:00 am. If I create a daily at 1:00 am and it starts the same day, it'll cause damage on the next cron in two hours (as I am unlikely to have done it). If it starts the next day, it won't show up as due until the next day *at midnight* when I'm (hopefully) fast asleep, and then cause damage on the cron. Either way, the daily only becomes active at the 'last minute' (i.e. within the last three hours before cron).

A couple of possible ideas: 
* Warn users that dailies created before CDR on the same day will cause damage (perhaps via notification?)
* Switch to `startDate` become a date & time, rather than a date. (Maybe make a new `startTimestamp` and calculate `startDate` on the fly? Or rename `startDate` to `startTimestamp` and warn of upcoming breaking changes)

**Comment by natfarleydev** [comment]
I've had a go at this, see #10783.
